lib/Auth/Process/PerunEntitlement.php

```diff
@@ -0,0 +1,201 @@
+            );
+        }
+        $this->forwardedEduPersonEntitlement = $config[self::FORWARDED_EDU_PERSON_ENTITLEMENT];
+
+        $this->groupNamePrefix = $conf->getString(self::GROUPNAMEPREFIX_ATTR, '');
+        $this->groupNameAuthority = $conf->getString(self::GROUPNAMEAUTHORITY_ATTR, '');
+        $this->groupNameAARC = $conf->getBoolean(self::GROUPNAMEAARC_ATTR, false);
+
+        if ($this->groupNameAARC && (empty($this->groupNameAuthority) || empty($this->groupNamePrefix))) {
+            throw new Exception(
+                'perun:PerunEntitlement: \'groupNameAARC\' has been set, \'groupNameAuthority\' ' .
+                'and \'groupNamePrefix\' options must be set as well'
+            );
+        }
+
+        if (!isset($config[self::INTERFACE_PROPNAME])) {
+            $config[self::INTERFACE_PROPNAME] = Adapter::RPC;
+        }
+
+        $this->interface = (string)$config[self::INTERFACE_PROPNAME];
+        $this->adapter = Adapter::getInstance($this->interface);
+    }
+
+    public function process(&$request)
+    {
+        $eduPersonEntitlement = $this->getEduPersonEntitlement($request);
+        $forwardedEduPersonEntitlement = $this->getForwardedEduPersonEntitlement($request);
+        $resourceCapabilities = $this->getResourceCapabilities($request);
+
+        $request['Attributes'][$this->eduPersonEntitlement] = array_merge(
+            $eduPersonEntitlement,
+            $forwardedEduPersonEntitlement,
+            $resourceCapabilities
+        );
+    }
+
+    private function getEduPersonEntitlement(&$request)
+    {
+        if (isset($request['perun']['groups'])) {
+            /** allow IDE hint whisperer
+             * @var model\Group[] $groups
+             */
+            $groups = $request['perun']['groups'];
+        } else {
+            throw new Exception(
+                'perun:PerunEntitlement: ' .
+                'missing mandatory field \'perun.groups\' in request.' .
+                'Hint: Did you configured PerunIdentity filter before this filter?'
+            );
+        }
+
+        $eduPersonEntitlement = [];
+        foreach ($groups as $group) {
+            if (isset($request['SPMetadata']['groupNameAARC']) || $this->groupNameAARC) {
+                # https://aarc-project.eu/wp-content/uploads/2017/11/AARC-JRA1.4A-201710.pdf
+                # Group name is URL encoded by RFC 3986 (http://www.ietf.org/rfc/rfc3986.txt)
+                # Example:
+                # urn:geant:einfra.cesnet.cz:perun.cesnet.cz:group:einfra:<groupName>:<subGroupName>#perun.cesnet.cz
+                if (empty($this->groupNameAuthority) || empty($this->groupNamePrefix)) {
+                    throw new Exception(
+                        'perun:PerunEntitlement: missing mandatory configuration options ' .
+                        '\'groupNameAuthority\' or \'groupNamePrefix\'.'
+                    );
+                }
+
+                $groupName = $this->groupNamePrefix . implode(
```

Comment:
Introduce private function that takes care of wrapping the value with prefix and suffix (also the encoding). Is it in all places where you do the wrapping.